### ISSUE TITLE
Email Showing as a Task and Task Form - Not Email and Email Form.

### ISSUE THREAD

**Issue by inspired27** [deejay-hub/timeline-lwc#285]
**Time Warp version**
Ross 1.14.0

**Describe the bug**
Email is showing as a task and uses task form not email form

**To reproduce**
Hover over Email or open link to detail from email on Time warp

**Expected behaviour**
When hovering over an email see the email details and when open link from email on Time Warp the email form opens

**Screenshots**
Included screenshots
Hover Over - got status and due date like a task but it is an email?:
![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/84cd3560-09b1-403b-8b31-8998537ab163)
Detail - Shows as Task using task form:
![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/d6c89e07-7f81-4599-9cbe-41eb1bcda12a)

This is what the actual email looks like from the built in timeline view the assumption is that the Time Warp should look like this also:
![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/626c02b0-8212-436a-a10a-6b052d431e43)

**Desktop (please complete the following information):**

Windows 10
Chrome (all versions)

**Additional context**
Enhanced Email Enabled - "Enhanced Email elevates email to a standard Salesforce object, with customization options and an API. Emails are no longer stored as tasks, but as full-featured records, complete with details, related lists, and collaboration tools. On email records, users can view their messages exactly as their customers do, complete with formatting."

**Comment by deejay-hub** [comment]
Hi @inspired27. Technically the emails are stored as Tasks _and_ EmailMessages when you turn on Enhanced Email - which creates 2 records for every email. The standard Timeline just ignores the Task. You can query or report on the Task - hidden but available.

The key to only show EmailMessages for Time Warp is 2-fold.

**1. You need to add the EmailMessage object to the Parent object. (In this example I'll use Contact but you can replicate for Account etc)**

a) Navigate to Setup->Custom Metadata Types
b) Click Manage Records near the record for Timeline Configuration

<img width="606" alt="image" src="https://github.com/deejay-hub/timeline-lwc/assets/7021883/231c473b-c118-44af-b6bb-1606e2c4c3ef">

c) Create a New record with the following values

<img width="1206" alt="image" src="https://github.com/deejay-hub/timeline-lwc/assets/7021883/d98a4dc5-f785-4ea6-99c2-4bedd3373c8f">

Now when I navigate to a Contact record it will plot the EmailMessage records. These will have a golden colour instead of grey based on the setup above.

Note that at this point you might see the Task AND the EmailMessage corresponding to the same email.

**2. You need to filter out Tasks corresponding to emails from Time Warp.**
You can either deactivate Tasks from showing by finding the metadata record for Contact_Tasks and unchecking the Active flag OR
You can filter out tasks that correspond to emails. See config guide in the section Filtering Records where we walk through how to remove Tasks.
https://quip.com/ahFCA9VBEKtr

**Comment by inspired27** [comment]
> Hi @inspired27. Technically the emails are stored as Tasks _and_ EmailMessages when you turn on Enhanced Email - which creates 2 records for every email. The standard Timeline just ignores the Task. You can query or report on the Task - hidden but available.
> 
> The key to only show EmailMessages for Time Warp is 2-fold.
> 
> **1. You need to add the EmailMessage object to the Parent object. (In this example I'll use Contact but you can replicate for Account etc)**
> 
> a) Navigate to Setup->Custom Metadata Types b) Click Manage Records near the record for Timeline Configuration
> 
> <img alt="image" width="606" src="https://private-user-images.githubusercontent.com/7021883/319138769-231c473b-c118-44af-b6bb-1606e2c4c3ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNDE2NDcsIm5iZiI6MTcxMjE0MTM0NywicGF0aCI6Ii83MDIxODgzLzMxOTEzODc2OS0yMzFjNDczYi1jMTE4LTQ0YWYtYjZiYi0xNjA2ZTJjNGMzZWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDNUMTA0OTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2ZmMjA3YTdhNDE4MDcyMjA4ZjJiOTI2NjMxNTIzNDkzMTUzZTM0NzYwNjVjMWMwMDQ3ODE1ZjUwOGQ1MTdhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.KKYG3qwpqEfunaagc0wTXVGQiUnmt3yxqZV9ExuY3Uc">
> c) Create a New record with the following values
> 
> <img alt="image" width="1325" src="https://private-user-images.githubusercontent.com/7021883/319146932-28130288-a1e6-4c2b-ad20-677d6b170530.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNDE2NDcsIm5iZiI6MTcxMjE0MTM0NywicGF0aCI6Ii83MDIxODgzLzMxOTE0NjkzMi0yODEzMDI4OC1hMWU2LTRjMmItYWQyMC02NzdkNmIxNzA1MzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDNUMTA0OTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTc0NzRmOGRhYjBhMTM0MGZiMDcwYmUyZmU3MjMwYWM2ZWRlMDExNGE5YmY0NWUxYzcxNmFkNTFlNTQ3ZWZhNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.NJP536nZO6THiA0OOIUt50l4Db4ahi6ikvAWq9iyDMs">
> Now when I navigate to a Contact record it will plot the EmailMessage records. These will have a golden colour instead of grey based on the setup above.
> 
> Note that at this point you might see the Task AND the EmailMessage corresponding to the same email.
> 
> **2. You need to filter out Tasks corresponding to emails from Time Warp.** You can either deactivate Tasks from showing by finding the metadata record for Contact_Tasks and unchecking the Active flag OR You can filter out tasks that correspond to emails. See config guide in the section Filtering Records where we walk through how to remove Tasks. https://quip.com/ahFCA9VBEKtr

I have followed the instructions above assume the icon is "/img/icon/t4v35/standard/email.svg" as in the screenshot it is cut off. However there is no change still the same outcome as I am getting:

![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/c949dadc-2d23-4756-9f4e-b06c2f379471)

![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/23c7bb34-e5be-4e75-afce-2cd01fc2b3d0)

![image](https://github.com/deejay-hub/timeline-lwc/assets/33864965/6b6b2cf2-268d-4188-8039-5f2096fe91b7)

**Comment by deejay-hub** [comment]
Yes. I see the task and the emailmessage record.

<img width="328" alt="image" src="https://github.com/deejay-hub/timeline-lwc/assets/7021883/d5d43592-38bf-4d7e-b360-e44654ad1cde">

I have configured for Contact only. I amended the setup slightly from my earlier screenshot and made sure I logged out and back in again since the data is cached.

<img width="1242" alt="image" src="https://github.com/deejay-hub/timeline-lwc/assets/7021883/3e5666fb-088e-44b2-a930-6d9f55185fe3">

**Comment by deejay-hub** [comment]
EmailMessage relationships as far as I know are not supported on the Account object. So your metadata record for the Account object is ignored since the relationship doesn't exist.

To show indirect relationships (e.g. to see emails sent to Contacts of the Account) to replicate the standard Timeline you have to use ActivityHistory and OpenActivity relationships for Account.

This will plot the Task version of the email and not the EmailMessage. So whilst it shows the Email record correctly the drilldown behaviour will be to the Task.

To see the behaviour you need to inactivate the Account_Tasks metadata record

![image](https://github.com/deejay-hub/timeline-lwc/assets/7021883/72677be1-8ba4-4dcc-846a-802c2a5f0416)

Then activate the records for Account_ActivityHistory and Account_OpenActivities

![image](https://github.com/deejay-hub/timeline-lwc/assets/7021883/1ac8fd2f-afb1-4cba-8519-6c1235671540)

![image](https://github.com/deejay-hub/timeline-lwc/assets/7021883/4af0d388-718d-44b3-b976-c5e558b7325a)

We don't support the ability to drilldown to the EmailMessage from the Account or to show fields from the EmailMessage record.

**Comment by deejay-hub** [comment]
We'll keep this open to look at how we can use the AlternateDetailId field on the ActivityHistory object to automatically drilldown to the EmailMessage for these use cases.

**Comment by deejay-hub** [comment]
Change logic to include AlternateDetailId is queries to ActivityHistory and OpenActivities.

When the AlternateDetailId is present then use the value as the recordId to drilldown to.

TimelineService 
```
if (
    (objName == 'OpenActivity' || objName == 'ActivityHistory') &&
    !selectStatement.contains('AlternateDetailId')
) {
    selectStatement = selectStatement + ', AlternateDetailId' + '';
}

private class TimelineRecord {
    private String alternateDetailId;
}
```

timeline.js
.on('click', function (event, d) {
  let drilldownId = d.recordId;
  if (d.alternateDetailId !== '') {
      drilldownId = d.alternateDetailId;
  }

**Comment by deejay-hub** [comment]
Might be worth clarifying the change.

1. Can I plot EmailMessage records on a Contact and Case - yes. Using the technique above. 

2. Can I plot EmailMessage records on an Account - no (not directly). This is because the Email records are indirectly related to the Account - the emails are linked to the Contact that is in turn linked to the Account. What we propose to do is show the Task version of the EmailMessage record (since they are a pair) using the ActivityHistory and OpenActivity relationship (since this includes the indirect records) then when someone clicks the 'Task' record corresponding to the EmailMessage it navigates them to the EmailMessage. From a users perspective it's seamless. They won't even notice it was the Task.

This change will be applied to the code base tomorrow in the epsilon branch and this issue will be flagged as closed.

As for the managed package on AppExchange. As it's a free, community managed package we don't commit to dates since we do this in our spare time. We loosely aim for 2 windows June and November. If it's a deal breaker you might need to find an alternative approach or package up a version from the source code.

Thanks for sharing the issue. Sorry we can't be more specific on timings.

**Comment by deejay-hub** [comment]
The source code is right here. So if you're familiar with the process you can take the code and put it into whichever environment you like directly. You can also release your own unmanaged OR managed package.

Each approach has pros and cons. But if you go this route you become the custodian of the code you downloaded so any changes have to be managed.

https://trailhead.salesforce.com/content/learn/trails/sfdx_get_started

**Comment by deejay-hub** [comment]
Closed in v1.15
